File: languages/korean/cpp/src/audio_processor.cpp
// src/cpp/src/audio_processor.cpp
#include "realtime_engine_ko/audio_processor.h"
#include "realtime_engine_ko/common.h"
#include <sstream>
#include <algorithm>
#include <cmath>
#include <chrono>
#include <thread>
#include <Eigen/Dense>
#include <numeric>

namespace realtime_engine_ko {

AudioProcessor::AudioProcessor(int sample_rate, float max_buffer_seconds)
    : sample_rate(sample_rate),
      max_buffer_length(static_cast<int>(sample_rate * max_buffer_seconds)),
      total_duration(0.0f) {
    
    std::stringstream ss;
    ss << "AudioProcessor 초기화: 샘플 레이트=" << sample_rate 
       << "Hz, 최대 버퍼 길이=" << max_buffer_length << " 샘플";
    LOG_INFO("AudioProcessor", ss.str());
}

AudioProcessor::~AudioProcessor() {
    // 정리 작업 (필요시)
}

Eigen::Matrix<float, Eigen::Dynamic, 1> AudioProcessor::ProcessAudioBinary(
    const std::vector<uint8_t>& binary_data) {
    
    try {
        // 바이너리 데이터 정보 출력
        LOG_INFO("AudioProcessor", "원본 바이너리 크기: " + std::to_string(binary_data.size()) + " 바이트");
        
        // 바이너리 데이터를 int16_t로 변환
        if (binary_data.size() % 2 != 0) {
            LOG_ERROR("AudioProcessor", "바이너리 데이터 크기가 올바르지 않습니다");
            return Eigen::Matrix<float, Eigen::Dynamic, 1>();
        }
        
        // int16을 float로 변환
        std::vector<float> audio_float(binary_data.size() / 2);
        const int16_t* int16_data = reinterpret_cast<const int16_t*>(binary_data.data());
        
        for (size_t i = 0; i < audio_float.size(); ++i) {
            audio_float[i] = static_cast<float>(int16_data[i]) / 32768.0f;
        }
        
        float expected_seconds = static_cast<float>(audio_float.size()) / sample_rate;
        LOG_INFO("AudioProcessor", "샘플 수: " + std::to_string(audio_float.size()) + 
                            ", 예상 시간: " + std::to_string(expected_seconds) + "초");
        
        // 버퍼 상태 출력
        LOG_INFO("AudioProcessor", "버퍼 추가 전: " + std::to_string(audio_buffer.size()) + 
                            " 샘플, 버퍼 시간: " + std::to_string(static_cast<float>(audio_buffer.size()) / sample_rate) + "초");
        
        // 오디오 버퍼에 추가
        audio_buffer.insert(audio_buffer.end(), audio_float.begin(), audio_float.end());
        
        // 최대 길이 제한
        if (audio_buffer.size() > max_buffer_length) {
            LOG_INFO("AudioProcessor", "오디오 버퍼 최대 길이에 도달했습니다. 길이: " + 
                std::to_string(audio_buffer.size()));
                
            // 버퍼 뒷부분(최신 데이터)만 유지
            audio_buffer.erase(audio_buffer.begin(), 
                audio_buffer.begin() + (audio_buffer.size() - max_buffer_length));
        }
        
        // 버퍼 통계
        float max_abs = 0.0f;
        for (const auto& sample : audio_buffer) {
            max_abs = std::max(max_abs, std::abs(sample));
        }
        
        LOG_INFO("AudioProcessor", "[CPP][AudioProcessor] 버퍼 통계: 최대값=" + std::to_string(max_abs) + 
                               ", 길이=" + std::to_string(audio_buffer.size()) + 
                               ", 시간=" + std::to_string(audio_buffer.size() / sample_rate) + "초");
        
        // VAD 직접 파라미터 지정 및 로깅
        const float vad_threshold = 0.00005f;
        const int min_frames = 10;  // Python과 일치시킴
        
        LOG_INFO("AudioProcessor", "[CPP][VAD] 파라미터: 임계값=" + std::to_string(vad_threshold) + 
                               ", 최소프레임=" + std::to_string(min_frames));
        
        bool has_voice = DetectVoiceActivity(audio_buffer, vad_threshold, min_frames);
        if (!has_voice) {
            LOG_INFO("AudioProcessor", "[CPP][AudioProcessor] VAD 실패: 음성 감지되지 않음");
            return Eigen::Matrix<float, Eigen::Dynamic, 1>();  // 빈 텐서 반환
        }
        
        // 전체 버퍼에 대한 전처리 및 반환
        auto result = PreprocessAudioData(audio_buffer);
        
        // 출력 텐서 통계
        if (result.size() > 0) {
            float max_val = 0.0f;
            for (int i = 0; i < result.size(); i++) {
                max_val = std::max(max_val, std::abs(result(i)));
            }
            LOG_INFO("AudioProcessor", "[CPP][AudioProcessor] 출력 텐서: 크기=" + std::to_string(result.size()) +
                                  ", 최대값=" + std::to_string(max_val));
        }
        
        return result;
    } catch (const std::exception& e) {
        std::stringstream ss;
        ss << "오디오 바이너리 처리 중 오류 발생: " << e.what();
        LOG_ERROR("AudioProcessor", ss.str());
        return Eigen::Matrix<float, Eigen::Dynamic, 1>();
    }
}

Eigen::Matrix<float, Eigen::Dynamic, 1> AudioProcessor::PreprocessAudioData(
    const std::vector<float>& audio_data, bool do_normalize) {
    
    // VAD 검사 - 음성이 없으면 빈 텐서 반환
    if (!DetectVoiceActivity(audio_data)) {
        return Eigen::Matrix<float, Eigen::Dynamic, 1>();
    }
    
    // 정규화
    std::vector<float> processed = audio_data;
    if (do_normalize) {
        // 평균 계산
        float mean = std::accumulate(audio_data.begin(), audio_data.end(), 0.0f) / audio_data.size();
        
        // 표준편차 계산
        float var = 0.0f;
        for (const auto& sample : audio_data) {
            var += (sample - mean) * (sample - mean);
        }
        float stddev = std::sqrt(var / audio_data.size()) + 1e-8f;
        
        // 정규화 적용
        for (size_t i = 0; i < processed.size(); ++i) {
            processed[i] = (processed[i] - mean) / stddev;
        }
    }
    
    // Eigen::Matrix로 변환
    Eigen::Matrix<float, Eigen::Dynamic, 1> result(processed.size());
    for (size_t i = 0; i < processed.size(); ++i) {
        result(i) = processed[i];
    }
    
    // 처리 통계 업데이트
    last_process_time = std::chrono::system_clock::now();
    total_duration = static_cast<float>(audio_buffer.size()) / static_cast<float>(sample_rate);
    
    return result;
}

bool AudioProcessor::DetectVoiceActivity(const std::vector<float>& audio_data, 
                                      float energy_threshold, int min_speech_frames) {
    // 프레임 단위로 분할 (10ms 프레임)
    int frame_size = static_cast<int>(sample_rate * 0.01);
    std::vector<std::vector<float>> frames;
    
    for (size_t i = 0; i < audio_data.size(); i += frame_size) {
        size_t end = std::min(i + frame_size, audio_data.size());
        if (end - i == frame_size) {  // 완전한 프레임만 고려
            frames.push_back(std::vector<float>(audio_data.begin() + i, audio_data.begin() + end));
        }
    }
    
    // 각 프레임의 에너지 계산
    std::vector<float> energies;
    for (const auto& frame : frames) {
        float energy = 0.0f;
        for (float sample : frame) {
            energy += sample * sample;
        }
        energy /= frame.size();
        energies.push_back(energy);
    }
    
    // 임계값을 넘는 프레임 수 계산
    int speech_frames = 0;
    for (float energy : energies) {
        if (energy > energy_threshold) {
            speech_frames++;
        }
    }
    
    // 평균 에너지 계산
    float avg_energy = 0.0f;
    if (!energies.empty()) {
        avg_energy = std::accumulate(energies.begin(), energies.end(), 0.0f) / energies.size();
    }
    
    // 최대 에너지 계산 추가
    float max_energy = 0.0f;
    if (!energies.empty()) {
        max_energy = *std::max_element(energies.begin(), energies.end());
    }
    
    LOG_INFO("AudioProcessor", "[CPP][VAD] 에너지 통계: 최대=" + std::to_string(max_energy) +
                          ", 평균=" + std::to_string(avg_energy) + 
                          ", 음성 프레임=" + std::to_string(speech_frames) + 
                          "/" + std::to_string(energies.size()));
    
    // 임계값을 넘는 프레임이 충분한지 확인
    return speech_frames >= min_speech_frames;
}

void AudioProcessor::Reset() {
    audio_buffer.clear();
    total_duration = 0.0f;
    last_process_time = std::chrono::system_clock::time_point();
    
    LOG_INFO("AudioProcessor", "상태 초기화 완료");
}

} // namespace realtime_engine_ko

File: languages/korean/python/realtime_engine_ko/audio_processor.py
import os
import time
import numpy as np
import soundfile as sf
import torch
from typing import Optional, Tuple, Dict, Any, List
import threading
import logging

# 로깅 설정
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
logger = logging.getLogger("AudioProcessor")

class AudioProcessor:
    """
    오디오 데이터 처리를 담당하는 클래스 (바이너리 입력 지원)
    """
    
    def __init__(
        self,
        sample_rate: int = 16000,
        max_buffer_seconds: float = 10.0
    ):
        """
        오디오 프로세서 초기화
        
        Args:
            sample_rate: 목표 샘플링 레이트 (Hz)
            max_buffer_seconds: 최대 버퍼 길이 (초)
        """
        self.sample_rate = sample_rate
        self.max_buffer_length = int(sample_rate * max_buffer_seconds)
        self.audio_buffer = np.array([], dtype=np.float32)
        
        # 처리 상태 추적
        self.last_process_time: Optional[float] = None
        self.total_duration: float = 0.0
        self.latest_chunk: Optional[torch.Tensor] = None
        
    def process_audio_binary(self, binary_data) -> Optional[torch.Tensor]:
        """
        바이너리 오디오 데이터 처리
        
        Args:
            binary_data: 바이너리 오디오 데이터 (일반적으로 int16 형식)
            
        Returns:
            Optional[torch.Tensor]: 처리된 오디오 텐서 또는 None
        """
        try:
            # 먼저 int16으로 가정하고 처리 (PyAudio 기본 포맷)
            audio_data = np.frombuffer(binary_data, dtype=np.int16).astype(np.float32) / 32768.0
            
            # 버퍼에 추가
            self.audio_buffer = np.append(self.audio_buffer, audio_data)
            
            # 최대 길이 제한
            if len(self.audio_buffer) > self.max_buffer_length:
                self.audio_buffer = self.audio_buffer[-self.max_buffer_length:]
            
            # 전체 버퍼에 대한 전처리 및 반환
            return self._preprocess_audio_data(self.audio_buffer)
            
        except Exception as e:
            logger.error(f"오디오 바이너리 처리 중 오류 발생: {e}")
            return None
    
    def _preprocess_audio_data(self, audio_data: np.ndarray, do_normalize: bool = True) -> Optional[torch.Tensor]:
        """
        오디오 데이터 직접 전처리
        
        Args:
            audio_data: numpy 배열 형태의 오디오 데이터
            do_normalize: 정규화 여부
            
        Returns:
            Optional[torch.Tensor]: 전처리된 오디오 텐서, 유효하지 않으면 None
        """
        # VAD 검사 - 음성이 없으면 None 반환
        if not self._detect_voice_activity(audio_data):
            return None
            
        # 1) 모노화 (이미 모노인 경우 건너뜀)
        if audio_data.ndim > 1:
            audio_data = np.mean(audio_data, axis=1)
        
        # 2) 정규화
        if do_normalize:
            m = audio_data.mean()
            s = audio_data.std()
            audio_data = (audio_data - m) / (s + 1e-8)
        
        # 3) float32로 캐스팅
        audio_data = audio_data.astype(np.float32)
        
        # 4) torch.Tensor로 변환 및 배치 차원 추가
        tensor = torch.from_numpy(audio_data).unsqueeze(0)  # shape: [1, T]
        
        return tensor
    
    def _detect_voice_activity(self, audio_data: np.ndarray, 
                            energy_threshold: float = 0.0005,
                            min_speech_frames: int = 10) -> bool:
        """
        간단한 에너지 기반 VAD 구현
        
        Args:
            audio_data: 오디오 데이터 (numpy 배열)
            energy_threshold: 음성으로 판단할 에너지 임계값
            min_speech_frames: 음성으로 판단할 최소 프레임 수
            
        Returns:
            bool: 음성이 있는지 여부
        """
        # 모노 데이터로 변환
        if audio_data.ndim > 1:
            audio_data = np.mean(audio_data, axis=1)
            
        # 프레임 단위로 분할 (10ms 프레임)
        frame_size = int(self.sample_rate * 0.01)
        frames = [audio_data[i:i+frame_size] for i in range(0, len(audio_data), frame_size)]
        
        # 각 프레임의 에너지 계산
        energies = [np.sum(frame**2) / len(frame) for frame in frames if len(frame) == frame_size]
        
        # 임계값을 넘는 프레임 수 계산
        speech_frames = sum(1 for energy in energies if energy > energy_threshold)
        
        # 로깅 (디버깅용)
        avg_energy = np.mean(energies) if energies else 0
        logger.debug(f"VAD: 평균 에너지={avg_energy:.6f}, 음성 프레임={speech_frames}/{len(energies)}")
        
        # 임계값을 넘는 프레임이 충분한지 확인
        return speech_frames >= min_speech_frames
    
    def reset(self) -> None:
        """상태 초기화"""
        self.total_duration = 0.0
        self.last_process_time = None
        self.latest_chunk = None
